cusub_perception/localizer/src/z_plane.cpp

```diff
@@ -0,0 +1,175 @@
+
+                    camera_matrix = occam_camera_matrix;
+                    dist_coefs = occam_dist_coefs;
+
+                } else if(image.header.frame_id.find("downcam") != string::npos){
+
+                    camera_matrix = downcam_camera_matrix;
+                    dist_coefs = downcam_dist_coefs;
+
+                } else {
+
+                    // Don't know what camera parameters are so we give up
+                    continue;
+
+                }
+
+                // Undistort the point
+                undistortPoints(points, undistorted_points, camera_matrix, dist_coefs, noArray(),  camera_matrix);
+
+                //ROS_DEBUG_STREAM("\nundistorted_points: \n" << undistorted_points);
+
+                // Project the point into camera space
+                vector<double> ray_points{
+                    undistorted_points[0].x, undistorted_points[0].y, 1
+                };
+                Mat ray_point{1,3,DataType<double>::type, ray_points.data()};
+
+                //ROS_DEBUG_STREAM("\nrp: \n" << ray_point);
+                //ROS_DEBUG_STREAM("\ninv: \n" << camera_matrix.inv());
+                //ROS_DEBUG_STREAM("\nt: \n" << ray_point.t());
+
+                Mat point_cam_space = camera_matrix.inv() * ray_point.t();
+
+                //ROS_DEBUG_STREAM("\npoint_cam_space: \n" << point_cam_space);
+
+                // Transform the point into world space as well as the camera center
+                std::string reference_frame("leviathan/description/odom");
+                std::string camera_frame(image.header.frame_id);
+
+                listener.waitForTransform(reference_frame, camera_frame, image.header.stamp, ros::Duration(5.0));
+
+                geometry_msgs::Point cam_pt =
+                    transformPoint(
+                        image.header,
+                        reference_frame,
+                        0,
+                        0,
+                        0
+                    );
+
+                geometry_msgs::Point ray_pt =
+                    transformPoint(
+                        image.header,
+                        reference_frame,
+                        point_cam_space.at<double>(0),
+                        point_cam_space.at<double>(1),
+                        point_cam_space.at<double>(2)
+                    );
+
+                //ROS_DEBUG_STREAM("\ncam: \n" << cam_pt);
+                //ROS_DEBUG_STREAM("\nraw: \n" << ray_pt);
+
+                // TODO check that the ray is not being projected at to shallow
+                // of an angle as that is likely to be very error prone
+
+                // Use the ray defined by the camera center and the ray point to
+                //  intersect the Z-Plane that is defined by the floor
+                //  We can do this by scaling the vector by the factor that makes
+                //  the z of the ray equal to the object height
+                double object_z;
+                ros::NodeHandle nh;
```

Comment:
I feel like the nh should be a member variable and initialized in the onInit(), because it may be slow to create one, but I could be wrong

Reply:
Probably, I only did it this way because its how it is in the other localizers constructors.  I guess there is also the consideration if we should have it check the parameter server every time.  Having it check every time makes it more like a dynamic reconfigure.  Could just load them all into a map on startup.

Reply:
Right, and in the future it could be something that changes dynamically, but it could easily be setup as a pub/sub rather than a dynamic reconfigure